File: 060_MinCostClimbingStairs.cpp
// Topic: Dynamic Programming
// Min Cost Climbing Stairs: dp with O(1) space keeping last two costs
// Time Complexity: O(n)
// Space Complexity: O(1)

#include <bits/stdc++.h>
using namespace std;

int minCostClimbingStairs(vector<int>& cost)
{
    int n = cost.size();
    int a = 0, b = 0;  // cost to reach i-2 and i-1
    for (int i = 2; i <= n; i++)
    {
        int takeOne = b + cost[i - 1];
        int takeTwo = a + cost[i - 2];
        int cur = min(takeOne, takeTwo);
        a = b;
        b = cur;
    }
    return b;
}

int main()
{
    cout << minCostClimbingStairs((vector<int>{10, 15, 20})) << "\n";                          // 15
    cout << minCostClimbingStairs((vector<int>{1, 100, 1, 1, 1, 100, 1, 1, 100, 1})) << "\n";  // 6
    return 0;
}

File: 060_MinCostClimbingStairs.py
"""
Topic: Dynamic Programming
Problem: Min Cost Climbing Stairs
Description: Minimum cost to reach the top using DP (either from step i or i+1).
Time: O(n)
Space: O(1)
"""

from typing import List


def minCostClimbingStairs(cost: List[int]) -> int:
    a = 0
    b = 0
    for x in reversed(cost):
        a, b = x + min(a, b), a
    return min(a, b)


if __name__ == '__main__':
    print(minCostClimbingStairs([10,15,20]))  # 15
    print(minCostClimbingStairs([1,100,1,1,1,100,1,1,100,1]))  # 6
